### PR TITLE
Split out monitor/det names into file info

### LINKED ISSUE #29534: SANS Move All Fileinformation fields into own state object
In various issues I've done I keep running into problems where some details, such as IPF / IDF or scaling factors, depend on the file information being loaded

This causes chicken and egg problems and requires us to even reload the entire state when a user inputs any files.
This made #28513 difficult, and is currently blocking #29490 

To fix, we should:
- Identify the fields loaded with file information (should be easy looking at the new interface)
- Move them into their own state object
- Update the algs to understand this new structure

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**

- Backports dataclasses, to remove boilerplate from adding new POD types
- Splits out monitor / detector names. These require reading from the IDF/IPF so conflating them into the state information (which is read from user files) causes chicken-egg problems on the GUI

**To test:**

- Check unit tests / system tests pass - this exercises all paths that uses State Objects
- Code review

Partially Fixes #29534 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 1]
```diff
@@ -9,7 +9,10 @@
 import math
 from mantid.api import MatrixWorkspace
 from abc import (ABCMeta, abstractmethod)
-from sans.state.StateObjects.StateMoveDetectors import StateMove
+
+from sans.state.AllStates import AllStates
+from sans.state.StateObjects.StateMoveDetectors import StateMove, StateMoveLOQ, StateMoveLARMOR, StateMoveZOOM, \
```
Comment (robertapplin) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py:14]: Its good practise to put these in alphabetical order - very picky I know!

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 2]
```diff
@@ -246,31 +248,31 @@ def set_selected_components_to_original_position(workspace, component_names):
...
 
 
-def set_components_to_original_for_isis(move_info, workspace, component):
+def set_components_to_original_for_isis(inst_info, workspace, component):
     """
     This function resets the components for ISIS instruments. These are normally HAB, LAB, the monitors and
     the sample holder
 
-    :param move_info: a StateMove object.
+    :param move_info: a StateInstrumentInfo object.
```
Comment (robertapplin) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py:256]: Should `move_info` be `inst_info` in this comment?

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 3]
```diff
@@ -333,53 +335,57 @@ def move_low_angle_bank_for_SANS2D_and_ZOOM(move_info, workspace, coordinates, u
...
+            for detector_keys in list(self.move_state.detectors.keys()):
+                is_name = component == detector_names[detector_keys].detector_name
+                is_name_short = component == detector_names[detector_keys].detector_name_short
                 if is_name or is_name_short:
                     found_name = True
                     break
             if not found_name:
                 raise ValueError("MoveInstrumentComponent: The component to be moved {0} cannot be found in the"
-                                 " state information of type {1}".format(str(component), str(type(move_info))))
+                                 " state information of type {1}".format(str(component), str(type(self.move_state))))
```
Comment (robertapplin) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py:388]: Could we update this to use an f-string?

File: scripts/SANS/sans/state/StateObjects/state_instrument_info.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+from sans.state.StateObjects.StateData import StateData
+
+
+@dataclass()
+class DetectorNames(metaclass=JsonSerializable):
+    detector_name: str = None
+    detector_name_short: str = None
+
+
+class StateInstrumentInfo(metaclass=JsonSerializable):
```
Comment (robertapplin) [inline scripts/SANS/sans/state/StateObjects/state_instrument_info.py:22]: Would it be useful to have some documentation on what this class is used for?

File: scripts/test/SANS/state/state_instrument_info_test.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+# SPDX - License - Identifier: GPL - 3.0 +
+import unittest
+from unittest import mock
+
+from sans.common.enums import SANSInstrument, DetectorType
+from sans.state.StateObjects import state_instrument_info
+from sans.state.StateObjects.state_instrument_info import StateInstrumentInfo
+
+
+class StateInstrumentInfoTest(unittest.TestCase):
```
Comment (robertapplin) [inline scripts/test/SANS/state/state_instrument_info_test.py:15]: Nice tests 👍

**Review by robertapplin — CHANGES_REQUESTED** [review]
Your changes look good, just a few minor comments/suggestions:

File: scripts/test/SANS/state/state_instrument_info_test.py (hunk 5 above)
Comment (DavidFair) [reply scripts/test/SANS/state/state_instrument_info_test.py:15]: Thanks :smiley:

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 1 above)
Comment (DavidFair) [reply scripts/SANS/sans/algorithm_detail/move_workspaces.py:14]: No it's a fair shout, it's a shame my IDE doesn't do this whilst optimising imports :+1:

**Comment by DavidFair** [comment]
Sorry for the extensive delay in addressing really useful feedback

**Review by robertapplin — APPROVED** [review]
Thanks for making those changes @DavidFair . Looks good 👍

**Comment by gemmaguest** [comment]
@DavidFair Does the linked issue need to be re-opened if this is a partial fix?

**Comment by DavidFair** [comment]
Yes, I'll make a mental note to use Fixes Partially instead in future PRs

**Pull request merged**